stumpy/gpu_stump.py

```diff
@@ -7,17 +7,110 @@
 import os
 
 import numpy as np
-from numba import cuda
+from numba import cuda, jit
 
 from . import core, config
 from .gpu_aamp import gpu_aamp
 
 logger = logging.getLogger(__name__)
 
 
+@jit  # equivalent to `__host__ __device__` in C++ CUDA
```

Comment:
The reason that I chose to change it to `@jit` is that device function can be called from (GPU) kernel function, and therefore, I couldn't find a way to do a unit test on it. (I tried to test it but I got an error regarding the fact that kernel is missing). So, after some search, I found t he following link:

https://developer.nvidia.com/blog/seven-things-numba/

In item 3, the author explains that we can use `@jit` to avoid duplication for unit testing cuda function. I also tested its performance and both `@cuda.jit` and `@jit` showed similar computing time.

Please let me know what you think about it.

Reply:
Just to confirm, we are talking about the test function here. I mean.... the wrapper kernel is going to be used for test only. right?

```
# wrapper kernel

@cuda.jit(...) # providing signature in the arguments 
def _gpu_searchsorted_kernel(a, v, side, bfs, nlevel):
    ...
   if side='left': call  `_gpu_search_sorted_left`
   if side='right': call `_gpu_search_sorted_right`
   ...
```

Something like this? And, then calling it in test function?

Or, do you think that we should add wrapper kernel for the performant version? I mean...the main kernel (`_compute_and_update_PI_kernel`) calls  `gpu_searchsorted_kernel` (but I do not how to allocate blocks/threads when one kernel calls another kernel)? (So, in this case, we may get rid of device function and put the `gpu_searchsorted` code directly inside gpu_searchsorted_kernel?)

I still need to think more but I just want to make sure that I understand your point of view correctly.

> I need to remember that the top-k NN will still be outside of the exclusion zone but that the top-k NN may not be spaced far away from each other.

Another thing that might be good to know: If, in the future, we decide to ignore trivial matches while searching top-k NN (or add it as an optional feature), I believe we can only do it in row-wise traversal. Because, we need to have all elements of distance profile first and then find the index with smallest distance, and then setting that index and its trivial neighbors to `inf` and search for the second one and so on. I think the diagonal traversal cannot be helpful in such case.

Reply:
> Firstly, I think it's perfectly fine for `_compute_and_update_PI_kernel` to call the device function `_gpu_searchsorted_right` directly. So, don't change anything here.

Sure.

> Maybe this is complicating things too much. | To be clear, no STUMPY function will actually call `_gpu_searchsorted` or `_gpu_searchsorted_kernel` (at least, not at the current moment) except for the unit test(s). Does that help?

So, how about the following approach? do you think it is less complicated?

```
# in tests/test_gpu_sorted 

from stumpy.gpu_stump import _gpu_searchsorted_left, _gpu_searchsorted_right

def gpu_searchsorted_kernel[1, 1](a, v, bfs, nlevel, is_left, idx):
    if is_left == 1:
        idx = _gpu_searchsorted_left(...)
   else:
        idx = _gpu_searchsorted_right(...) 


idx = -1
gpu_searchsorted_kernel(a, v, bfs, nlevel, is_left, idx)
# asserting if idx == np.searchsorted(a, v, side=...)
```

We can use a for-loop to test it on different test inputs. What do you think?

Reply:
I think you might be missing the point. During the `unit test` phase of `test.sh`, when a GPU is present, the goal should ALWAYS be to execute the test on the GPU device and NOT on the host (CPU). If no GPU is present then the tests are skipped. The problem is that we have a "device function" and this is new in STUMPY. We need a way to test out this device function in the context of a simple but real scenario. I don't think it's great to set the `blocks_per_grid` and `threads_per_block` to `1` as this seems unrealistic. I really don't know...

During the `coverage` phase of `test.sh`, CUDA is disabled and then the @cuda.jit` automatically "simulated". This implies that it is all completed on the "host" (CPU).

Reply:
I tested something like this and I think it might work:

```
from numba import cuda
import numpy as np
import math

cuda.is_available()

@cuda.jit(device=True)
def _gpu_searchsorted_left(a, v):
    # I faked something shorter here since I am testing on Google Colab
    # and I didn't have access to the _bfs_indices function
    for i in range(a.shape[0]):
        if v < a[i]:
            return i


@cuda.jit("(f8[:, :], f8[:], i8[:])")
def _gpu_searchsorted_kernel(A, V, idx):
    i = cuda.grid(1)

    if i < A.shape[0]:
        idx[i] = _gpu_searchsorted_left(A[i], V[i])
  

def test_gpu_searchsorted():
    n = 5000
    k = 20
    A = np.random.rand(n * k).reshape(n, k)
    V = np.random.rand(n)
    idx = np.full(n, -1, dtype=np.int64)
    threads_per_block = 100  # config.STUMPY_THREADS_PER_BLOCK
    blocks_per_grid = math.ceil(n / threads_per_block)
    _gpu_searchsorted_kernel[blocks_per_grid, threads_per_block](
        A, V, idx
    )
    print(idx)
  

test_gpu_searchsorted()
```

Reply:
> During the coverage phase of test.sh, CUDA is disabled and then the @cuda.jit` automatically "simulated". This implies that it is all completed on the "host" (CPU).

> During the `unit test` phase of `test.sh`, when a GPU is present, the goal should ALWAYS be to execute the test on the GPU device and NOT on the host (CPU).

I see. Thanks for letting me know about this! So, better NOT to go to the turning-off-CUDA approach, and test it on GPU.

>  I don't think it's great to set the blocks_per_grid and threads_per_block to 1 as this seems unrealistic. I really don't know...

I totally agree as it is not realistic. 

However, if we just focus on the testing of the `gpu_searchsorted_right/left` just to see if it gives the expected output, the setting `def gpu_searchsorted_kernel[1,1](...)` should not be that bad. Because, in STUMPY, the function `gpu_searchsorted_right`  is being called (by kernel) in each thread and performs `searchsort` on 1-d array. We can test 100 cases in parallel to each other using different setting other than `1`, but that just speed up the testing process and does not have any particular advantage compared to`def gpu_searchsorted_kernel[1,1](...)`. 

>  I don't think it's great to set the blocks_per_grid and threads_per_block to 1 as this seems unrealistic. I really don't know.

I have no idea :) I can do some more research to see if there is a clean/easy way to test a device function.

Reply:
@seanlaw 
Yeah... I noticed that :) I think we almost posted at the same time. 

> I tested something like this and I think it might work:

Got your point! So, I will create a wrapper kernel in `test_gpu_stump` and use it in test functions. Thanks for the suggestion!! I will try it out and let you know if things go wrong.